osu.Game/Online/RealtimeMultiplayer/StatefulMultiplayerClient.cs

```diff
@@ -0,0 +1,389 @@
+        private int playlistItemId; // Todo: THIS IS SUPER TEMPORARY!!
+
+        /// <summary>
+        /// Joins the <see cref="MultiplayerRoom"/> for a given API <see cref="Room"/>.
+        /// </summary>
+        /// <param name="room">The API <see cref="Room"/>.</param>
+        public async Task JoinRoom(Room room)
+        {
+            Debug.Assert(Room == null);
+            Debug.Assert(room.RoomID.Value != null);
+
+            apiRoom = room;
+            playlistItemId = room.Playlist.SingleOrDefault()?.ID ?? 0;
+
+            Room = await JoinRoom(room.RoomID.Value.Value);
+
+            Debug.Assert(Room != null);
+
+            foreach (var user in Room.Users)
+                await PopulateUser(user);
+
+            updateLocalRoomSettings(Room.Settings);
+        }
+
+        /// <summary>
+        /// Joins the <see cref="MultiplayerRoom"/> with a given ID.
+        /// </summary>
+        /// <param name="roomId">The room ID.</param>
+        /// <returns>The joined <see cref="MultiplayerRoom"/>.</returns>
+        protected abstract Task<MultiplayerRoom> JoinRoom(long roomId);
+
+        public virtual Task LeaveRoom()
+        {
+            if (Room == null)
+                return Task.CompletedTask;
+
+            apiRoom = null;
+            Room = null;
+
+            Schedule(() => RoomChanged?.Invoke());
+
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Change the current <see cref="MultiplayerRoom"/> settings.
+        /// </summary>
+        /// <remarks>
+        /// A room must be joined for this to have any effect.
+        /// </remarks>
+        /// <param name="name">The new room name, if any.</param>
+        /// <param name="item">The new room playlist item, if any.</param>
+        public void ChangeSettings(Optional<string> name = default, Optional<PlaylistItem> item = default)
+        {
+            if (Room == null)
+                return;
+
+            // A dummy playlist item filled with the current room settings (except mods).
+            var existingPlaylistItem = new PlaylistItem
+            {
+                Beatmap =
+                {
+                    Value = new BeatmapInfo
+                    {
+                        OnlineBeatmapID = Room.Settings.BeatmapID,
+                        MD5Hash = Room.Settings.BeatmapChecksum
+                    }
+                },
+                RulesetID = Room.Settings.RulesetID
+            };
+
+            var newSettings = new MultiplayerRoomSettings
+            {
+                Name = name.GetOr(Room.Settings.Name),
+                BeatmapID = item.GetOr(existingPlaylistItem).BeatmapID,
+                BeatmapChecksum = item.GetOr(existingPlaylistItem).Beatmap.Value.MD5Hash,
+                RulesetID = item.GetOr(existingPlaylistItem).RulesetID,
+                Mods = item.HasValue ? item.Value!.RequiredMods.Select(m => new APIMod(m)).ToList() : Room.Settings.Mods
```

Comment:
still of the opinion this operator should be banned outside of declarations

```suggestion
                Mods = item.HasValue ? item.Value.AsNonNull().RequiredMods.Select(m => new APIMod(m)).ToList() : Room.Settings.Mods
```

also slightly confused why the nullability annotations aren't working as I'd think they should (`Optional<T>.Value` returns `T`, which I wouldn't expect to be nullable - maybe that's an exception due to a lack of any type restrictions on `T` whatsoever?)

Reply:
Maybe because of default struct init :/